src/utils/pick.ts

```diff
@@ -0,0 +1,20 @@
+import { isArray } from './is-array';
+
+// NOTE: If changing these types, ensure the following cases cause a type error:
+// const noProps = pick({ a: 1 });
+// noProps.a; // Property 'a' does not exist on type 'Pick<{ a: number; }, never>'.
+// const emptyArr = pick({ a: 1 }, []);
+// emptyArr.a; // Property 'a' does not exist on type 'Pick<{ a: number; }, never>'.
+
+export function pick<T extends object, K extends keyof T>(obj: T, props: K[]): Pick<T, K>;
+export function pick<T extends object, K extends keyof T = never>(obj: T, ...props: K[]): Pick<T, K>;
+export function pick<T extends object, K extends keyof T>(obj: T, ...args: K[] | [ K[] ]): Pick<T, K> {
+   const props = ((args.length === 1 && isArray(args[0])) ? args[0] : args) as K[];
+
+   return props.reduce((memo, prop) => {
+      if (Object.prototype.hasOwnProperty.call(obj, prop)) {
```

Comment:
Just something to be aware of as a difference between this implementation and Underscore's: This implementation will not allow you to `pick` methods from class instances, but Underscore's does:

```js
class A {
   aMethod() {}
}

pick(new A(), 'aMethod'); // Our implementation returns an empty object: {}
_.pick(new A(), 'aMethod'); // Underscore returns:  { aMethod() {} }
```

Also, if the object has any properties set via a prototype chain, we can't get at them either:

```js
let p1 = { a: 'a' }

p1.__proto__ = { b: 'b' }

p1.b; // 'b'

pluck(p1, 'b'); // {}
_.pluck(p1, 'b'); // { b: 'b' }
```

Not that we would really do anything like that in our code, but it's possible third party libs might. I don't know that it's something we really have to worry about, I'm just pointing it out. I actually think checking `hasOwnProperty` and disallowing methods to be plucked is an improvement.

Should we mention anything about either of those cases in a doc block on the function?

Reply:
Good catch! In the most recent push I switched this to follow [underscore.pick's](https://github.com/jashkenas/underscore/blob/58df1085cdb05cb0888719c5fe5493948604ab69/modules/pick.js#L16) [use](https://github.com/jashkenas/underscore/blob/58df1085cdb05cb0888719c5fe5493948604ab69/modules/pick.js#L16) of the `in` operator.

I agree `hasOwnProperty` is much more specific. However, I couldn't find a way to get the types to exclude inherited properties. Thus, the types would say `pick(new A(), 'aMethod');` "works" when in reality it doesn't. Remembering this, even if there's a comment, adds to developer overhead. Seems to me that reducing the gotchas for the dev is the better route to go.

Reply:
Ah yeah, that makes sense. Even better. 👍🏻